Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Distinct.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT License.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Reaqtive.Operators
+{
+    internal sealed class Distinct<TResult, TKey> : SubscribableBase<TResult>
+    {
+        private readonly ISubscribable<TResult> _source;
+        private readonly Func<TResult, TKey> _keySelector;
+        private readonly IEqualityComparer<TKey> _equalityComparer;
+
+        public Distinct(ISubscribable<TResult> source, Func<TResult, TKey> keySelector, IEqualityComparer<TKey> comparer)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(keySelector != null);
+            Debug.Assert(comparer != null);
+
+            _source = source;
+            _keySelector = keySelector;
+            _equalityComparer = comparer;
+        }
+
+        protected override ISubscription SubscribeCore(IObserver<TResult> observer)
+        {
+            return new _(this, observer);
+        }
+
+        private sealed class _ : StatefulUnaryOperator<Distinct<TResult, TKey>, TResult>, IObserver<TResult>
+        {
+            private const string MAXDISTINCTELEMENTS = "rx://operators/distinct/settings/maxDistinctElements";
+            private int _maxDistinctElements;
+
+            private HashSet<TKey> _keySet;
+
+            public _(Distinct<TResult, TKey> parent, IObserver<TResult> observer)
+                : base(parent, observer)
+            {
+            }
+
+            public override void SetContext(IOperatorContext context)
+            {
+                base.SetContext(context);
+
+                context.TryGetInt32CheckGreaterThanZeroOrUseMaxValue(MAXDISTINCTELEMENTS, out _maxDistinctElements);
+            }
+
+            protected override void OnStart()
+            {
+                _keySet ??= new HashSet<TKey>(Params._equalityComparer);
+            }
+
+            public override string Name => "rc:Distinct";
+
+            public override Version Version => Versioning.v1;
+
+            public void OnCompleted()
+            {
+                Output.OnCompleted();
+                Dispose();
+            }
+
+            public void OnError(Exception error)
+            {
+                Output.OnError(error);
+                Dispose();
+            }
+
+            public void OnNext(TResult value)
+            {
+                var added = false;
+                try
+                {
+                    TKey key = Params._keySelector(value);
+
+                    // TODO: Should this operator silently stop propagating when it reaches max capacity? or should it throw exception?
```

Comment:
We should propagate an exception down the `OnError` channel, just like we do for other operators that have such a threshold. Otherwise, we end up silently producing incorrect sequences, which is semantically incorrect and unacceptable.

While erroring out is not ideal (cf. my discussion on operator local storage as the way out of this), it prevents state explosion in cases where it matters (i.e. places where users opt in). Also note that these settings come from the operator context, which can be controlled on a per-subscription basis.

Concrete deployments of Reaqtor have used this as a quota mechanism where different queries get different storage quota, while others chose other policies such as conservative global settings for entire engines, or simply don't deploy operators that have potentially unbounded storage requirements. Initially, we were leaning towards the latter by simply omitting such operators (e.g. `Buffer`, `Zip`, etc.) but as legitimate use cases grew (often on bounded sequences produced by e.g. a `Window` operator), we ended up adding more of these operators (with configurable quota settings) to the Reaqtive library.

One more piece of history is worth pointing out here. When we started this project, Reaqtive was an implementation detail of Reaqtor, and was only used in that context. As we shrunk down the footprint of engines and improved the experience on manually using the Reaqtive layer (with convenience subscription visitors used to initialize, load/save state, etc.), we decided to decouple of operator library from the engine more and more. As part of the OSS effort, we truly split off the Reaqtive libraries at their own entity in the "triad" of Nuqleon, Reaqtive, Reaqtor. (There was some complexity around this especially for higher-order operators, which bank on some facilities provided by the hosting environment.) This will allow for broader reuse.

Finally, note that the setting is optional and we'll default to `int.MaxValue`, which corresponds to the Rx default (i.e. there is no bound over there).

Reply:
See https://github.com/reaqtive/reaqtor/blob/main/Reaqtive/Core/Reaqtive.Linq/Reaqtive/Operators/Buffer.cs#L42 for the approach we take.